### ISSUE TITLE
Figure out how to get .m4b to save processing time

### ISSUE THREAD

**Issue by burntcookie90** [burntcookie90/librofm-downloader#8]

**Comment by drallgood** [comment]
I've been looking into this but haven't found an API yet.
The iOS app seems to not use it either (confirmed via Proxyman)

When using the web UI, there is a button to download it which then redirects to the correct signed asset request. However I have not figured out whether the API can do that too.

PS: Current version of the API seems to be v10

**Comment by burntcookie90** [comment]
Yeah, i poked around on web for awhile but the signed assets would be difficult to handle without API support. Poking through the android app doesn't show any API support for `.m4b` at all.

> PS: Current version of the API seems to be v10

let me make an issue to unify the endpoints. I grabbed them off some other libro fm code or issues.

**Comment by drallgood** [comment]
Heads up. I'm working on a way to at least convert the books to m4b in the meantime. 

Haven't figured out how to do this yet. Been poking and prodding the API and looking at their app's code but there doesn't appear to be any way to fetch the m4b instead 

Already got a working solution but ffmpeg is currently struggling with paths and special characters still. 
Will work that out later today and then submit a PR to get your feedback (while I do some cleanup and documentation)

**Comment by burntcookie90** [comment]
Oh cool, I look forward to it! Unfortunately, I'm pretty sure their mobile API doesn't have an endpoint for m4b downloading. It's gonna have to be some custom work in the service. 


One thing id thought of is looking to see if audiobookshelf could trigger an m4b conversion via its API, but then we'd have a hard dependency on it.

**Comment by drallgood** [comment]
Got ffmpeg to cooperate. Have a look at #51 

I chose ffmpeg instead of another dependency because that's the safer choice and also it works. 
I used a wrapper library I have used in previous projects but technically you could do without. Saw little risk there as it literally just makes life easier and if it ever were to become an issue (which I don't expect) would be easy enough to replace.

**Comment by burntcookie90** [comment]
I'm going to leave this open because I'd like it around in case libro.fm's API updates to serve M4B files.

**Comment by drallgood** [comment]
@burntcookie90 My thoughts exactly!

**Comment by burntcookie90** [comment]
looks like the api updated

`libro.fm/api/v10/audiobooks/{isbn}/packaged_m4b`

**Comment by burntcookie90** [comment]
@drallgood started implementing the API and it significantly cleans up the code. I know (and highly appreciate!) all the work that went into the ffmpeg solution but I think we can reduce complexity by removing it.

**Comment by drallgood** [comment]
@burntcookie90 Makes sense. I was wondering about that too. Only doubt I had was whether ALL books have an associated m4b file available?

**Comment by burntcookie90** [comment]
That's a very good point! I'll need to poke around and see what a failure case looks like.

**Comment by burntcookie90** [comment]
Ok, can confirm that the endpoint 404s if there isn't an m4b available. Will probably need to have a way to have ffmpeg based conversion as a fallback, and bare mp3 download as a fallback fallback.

example: https://libro.fm/api/v10/audiobooks/9781508287056/packaged_m4b

**Comment by drallgood** [comment]
> Ok, can confirm that the endpoint 404s if there isn't an m4b available. Will probably need to have a way to have ffmpeg based conversion as a fallback, and bare mp3 download as a fallback fallback.
> 
> example: https://libro.fm/api/v10/audiobooks/9781508287056/packaged_m4b

 Crap 😔 I was fearing that

**Comment by burntcookie90** [comment]
Wow, since I shared that 404 test case, it is now being served as an m4b. Going through my library and only a few dont have m4b now
